### ISSUE TITLE
Provides a component optionaly

### ISSUE THREAD

**Issue by a-peyrard** [restx/restx#213]
I just have the need of something like that:
```java
@Provides
Optional<AComponent> aComponentOptional() {
   return ...
}
```
I want to provide an object if and only if a file exists.

I think it could be a new feature for the restx-factory ? (not sure of the impacts though)

WDYT ? Is that something you once needed ?

**Comment by xhanin** [comment]
Sounds interesting, currently there is only the `@Alternative` with `@When` way to do that, but having a way to do it in a module sounds interesting too. The main difference is that with the `@When` only named strings are allowed to be provisionned to know if the component must be produced, its real dependencies are not provided. With the approach you suggest you will have to provide the dependencies even if in the end you don't provide the component. Not a big deal, just something to be aware of.

**Comment by fcamblor** [comment]
How will it behave with `priority` fields ?
That is to say, if I have :
```
@Module(priority=999)
public class Module1 {
  @Produces  Optional<String> blah(){ return Optional.absent(); }
}
@Module(priority=0)
public class Module2 {
  @Produces Optional<String> blah(){ return Optional.of("blah"); }
}
```

In that case, what will be produced for the `blah` component ?
An absent value (because priority takes precedence) or `Optional("blah")` because Module1.blah() has been evaluated and we saw it has been evaluated to an empty component, so we'll try to find another component with lower priority.

I don't really know which one to choose (the later may be more complicated to implement I think)

**Comment by a-peyrard** [comment]
@fcamblor Indeed that's an interesting question, note that your example could also been written like this  (in order to be a little bit more messy :)):
```java
@Module(priority=999)
public class Module1 {
  @Produces  Optional<String> blah(){ return Optional.absent(); }
}
@Module(priority=0)
public class Module2 {
  @Produces String blah(){ return "blah"; }
}
```
(And I guess, you meant a priority equals to -999 in Module1, in order to take precedence other Module2)

For me the goal would be to analyze all machine till one is producing something.

@xhanin yes dependencies will be built, but as you said, I don't think it would be a problem.

PS: I don't know if you saw it, but I also created another issue for alternatives a couple days ago: 
https://github.com/restx/restx/issues/212

**Comment by xhanin** [comment]
I agree with @a-peyrard I would go ahead until something is produced, if a module can do it. So producing absent is like having nothing, it won't prevent another machine from building a component for that name. If you really want to achieve that you need to produce a deactivation key.

**Comment by a-peyrard** [comment]
Just a quick thought before the WE, with this new feature, `NoopFactoryMachine` will become useless. `FactoryMachine` used for conditional components will just return `Optional.absent()` if condition is not fulfil.
